### PR TITLE
Add tests for hard links

### PR DESCRIPTION
**Pull request by lptr**

### CODE REVIEW

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+@Requires({ Platform.current().macOs || Platform.current().linux || Platform.current().windows })
+class HardLinkFileEventFunctionsTest extends AbstractFileEventFunctionsTest {
+
+    def "can detect hard link created"() {
+        given:
+        def target = new File(rootDir, "target")
+        target.createNewFile()
+        def link = new File(rootDir, "link")
+        startWatcher(rootDir)
```
Comment (wolfs) [inline file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:39]: Should we also have tests where the target of the hard link is outside of the watched directory?

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        when:
+        link.text = "modified"
+
+        then:
+        // On Windows we seem to be getting multiple MODIFIED events
+        // On Linux we _sometimes_ seem to be getting multiple events
+        expectEvents byPlatform(
+            (MAC_OS): [change(MODIFIED, link)],
+            (WINDOWS): [change(MODIFIED, link)] * 2,
```
Comment (wolfs) [inline file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:70]: I suppose that is one modified event for each of the watched files: one for link and one for target.

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        startWatcher(rootDir)
+
+        when:
+        link.text = "modified"
+
+        then:
+        // On Windows we seem to be getting multiple MODIFIED events
+        // On Linux we _sometimes_ seem to be getting multiple events
+        expectEvents byPlatform(
+            (MAC_OS): [change(MODIFIED, link)],
```
Comment (wolfs) [inline file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:69]: This and the linux behaviour is bad: If you modify the target outside of the watched directory, then we are not getting an event, right?

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy (hunk 1 above)
Comment (lptr) [reply file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:39]: Just added, it's not nice. :/

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy (hunk 2 above)
Comment (lptr) [reply file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:90]: Perhaps yes. We could try with more than two links, though whatever we'd learn doesn't sound useful.

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy (hunk 3 above)
Comment (lptr) [reply file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:89]: Actually, even on Windows you get no notification. :/

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+        then:
+        // Windows sometimes reports a modification before the removal
+        expectEvents byPlatform(
+            (WINDOWS): [optionalChange(MODIFIED, link), change(REMOVED, link)],
+            (OTHERWISE): [change(REMOVED, link)]
+        )
+    }
+
+    def "does not detect hard link removed outside watched hierarchy"() {
```
Comment (wolfs) [inline file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:166]: Same here: If I remove a link pointing to a file in the watched directory, why would that even matter? Is this just for coverage so we don't see any strange events?

**Review by wolfs — APPROVED** [review]
This looks "good", in the sense that it tests what is currently happening. At least on Linux, I would have expected to obtain the modification event, though seems like no dice...

File: file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy (hunk 4 above)
Comment (lptr) [reply file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy:166]: Exactly, just as above.

**Pull request merged**